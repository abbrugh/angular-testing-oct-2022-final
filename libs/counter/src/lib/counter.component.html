<div data-testid="counter-play">
  <button
    class="btn btn-danger"
    data-testid="decrement"
    (click)="decrement()">
    Decrement
  </button>
  <span data-testid="current">{{ current$ | async }}</span>
  <button
    (click)="increment()"
    data-testid="increment">
    Increment
  </button>
</div>

<router-outlet></router-outlet>
