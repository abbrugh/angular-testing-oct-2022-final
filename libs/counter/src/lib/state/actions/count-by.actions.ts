import { createActionGroup, props } from '@ngrx/store';
import { Options } from '../reducers/counter.reducer';

export const CountByEvents = createActionGroup({
  source: 'counter count-by events',
  events: {
    set: props<{ payload: Options }>(),
  },
});
