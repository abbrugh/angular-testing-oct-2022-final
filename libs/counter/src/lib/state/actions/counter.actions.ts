import { createActionGroup, emptyProps, props } from '@ngrx/store';
import { CountState } from '../reducers/counter.reducer';

export const CountEvents = createActionGroup({
  source: 'counter count events',
  events: {
    increment: emptyProps(),
    decrement: emptyProps(),
  },
});

export const CountDocuments = createActionGroup({
  source: 'counter count documents',
  events: {
    state: props<{ payload: CountState }>(),
  },
});

export const FeatureEvents = createActionGroup({
  source: 'counter feature events',
  events: {
    entered: emptyProps(),
  },
});
