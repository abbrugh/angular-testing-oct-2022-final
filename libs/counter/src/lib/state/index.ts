import {
  ActionReducerMap,
  createFeatureSelector,
  createSelector,
} from '@ngrx/store';

import * as fromCounter from './reducers/counter.reducer';

export const featureName = 'counter';

export interface CounterState {
  count: fromCounter.CountState;
}

export const reducers: ActionReducerMap<CounterState> = {
  count: fromCounter.reducers,
};

const selectFeature = createFeatureSelector<CounterState>(featureName);

export const selectCountBranch = createSelector(selectFeature, (f) => f.count);

export const selectCurrentCount = createSelector(
  selectCountBranch,
  (b) => b.current,
);

export const selectCountingBy = createSelector(selectCountBranch, (b) => b.by);
