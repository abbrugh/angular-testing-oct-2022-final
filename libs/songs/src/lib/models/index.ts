import { SongListEntity } from '../state/reducers/list.reducer';

export type SongListItemModel = SongListEntity;

export type PlaylistCreateModel = Omit<SongListEntity, 'id'>;
