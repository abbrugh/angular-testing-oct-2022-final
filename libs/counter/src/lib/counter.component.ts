import { Component, OnInit } from '@angular/core';
import { Store } from '@ngrx/store';
import { selectCurrentCount } from './state';
import { CountEvents, FeatureEvents } from './state/actions/counter.actions';

@Component({
  selector: 'ht-counter',
  templateUrl: './counter.component.html',
  styleUrls: ['./counter.component.css'],
})
export class CounterComponent {
  current$ = this.store.select(selectCurrentCount);

  constructor(private store: Store) {
    store.dispatch(FeatureEvents.entered());
  }
  increment() {
    this.store.dispatch(CountEvents.increment());
  }

  decrement() {
    this.store.dispatch(CountEvents.decrement());
  }
}
