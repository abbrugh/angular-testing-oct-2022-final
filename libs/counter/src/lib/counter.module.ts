import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CounterComponent } from './counter.component';
import { RouterModule, Routes } from '@angular/router';
import { CountByComponent } from './components/count-by/count-by.component';
import { StoreModule } from '@ngrx/store';
import { featureName, reducers } from './state';
import { CounterEffects } from './state/effects/counter.effects';
import { EffectsModule } from '@ngrx/effects';
import { CounterAdapter } from './servicers/counter.adapter';

const routes: Routes = [
  {
    path: '',
    component: CounterComponent,
    children: [
      {
        path: 'by',
        component: CountByComponent,
      },
    ],
  },
];

const counterToUse = new CounterAdapter('counter-key');
@NgModule({
  imports: [
    CommonModule,
    RouterModule.forChild(routes),
    StoreModule.forFeature(featureName, reducers),
    EffectsModule.forFeature([CounterEffects]),
  ],
  declarations: [
    CounterComponent,
    CountByComponent,
  ],
  providers: [{ provide: CounterAdapter, useValue: counterToUse }],
})
export class CounterModule {}
