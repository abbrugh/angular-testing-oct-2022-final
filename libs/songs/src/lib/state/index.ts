import {
  ActionReducerMap,
  createFeatureSelector,
  createSelector,
} from '@ngrx/store';
import { SongListItemModel } from '../models';
import * as fromList from './reducers/list.reducer';
export const featureName = 'songsList';

export interface SongsState {
  list: fromList.SongListState;
}

export const reducers: ActionReducerMap<SongsState> = {
  list: fromList.reducer,
};

const selectFeature = createFeatureSelector<SongsState>(featureName);

const selectListBranch = createSelector(selectFeature, (f) => f.list);

const { selectAll: selectSongListEntityArray } =
  fromList.adapter.getSelectors(selectListBranch);

export const selectSongListModel = createSelector(
  selectSongListEntityArray,
  (songs) => songs as SongListItemModel[],
);
