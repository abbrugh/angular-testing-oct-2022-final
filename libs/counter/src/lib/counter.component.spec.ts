import { CounterComponent } from './counter.component';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { provideMockStore, MockStore } from '@ngrx/store/testing';
import * as fromCounterReducer from '../lib/state/';
import { RouterModule } from '@angular/router';
import { CountEvents } from './state/actions/counter.actions';
describe('CounterComponent', () => {
  let spectator: Spectator<CounterComponent>;
  let store: MockStore;
  const initialState = {
    counter: {
      count: fromCounterReducer.reducers,
    },
  };
  const createComponent = createComponentFactory({
    component: CounterComponent,
    providers: [
      provideMockStore({
        initialState,
        selectors: [
          {
            selector: fromCounterReducer.selectCurrentCount,
            value: 99,
          },
        ],
      }),
    ],
    imports: [RouterModule],
  });

  beforeEach(() => {
    spectator = createComponent();
    store = spectator.inject(MockStore);
    spectator.detectChanges();
  });
  it('initial state', () => {
    const component = spectator.query('[data-testid="counter-play"]');
    expect(component).toMatchSnapshot();
  });
  it('can increment', () => {
    const incrementButton = spectator.query('[data-testid="increment"]');
    if (!incrementButton) {
      fail('Selector for increment button is bad.');
    } else {
      jest.spyOn(store, 'dispatch');
      spectator.click(incrementButton);
      expect(store.dispatch).toHaveBeenCalledWith(CountEvents.increment());
    }
  });
});
