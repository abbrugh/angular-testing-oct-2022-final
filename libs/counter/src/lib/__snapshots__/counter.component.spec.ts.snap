// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterComponent initial state 1`] = `
<div
  data-testid="counter-play"
>
  <button
    class="btn btn-danger"
    data-testid="decrement"
  >
     Decrement 
  </button>
  <span
    data-testid="current"
  >
    99
  </span>
  <button
    data-testid="increment"
  >
     Increment 
  </button>
</div>
`;
